### ISSUE TITLE
[Proposal] Add support to parameter serialization rules

### ISSUE THREAD

**Issue by alephao** [airtasker/spot#406]
## Introduction
There is no way to specify parameter serialization rules defined by openapi:

https://swagger.io/docs/specification/serialization/#query
http://spec.openapis.org/oas/v3.0.2#parameterExplode

## Proposed solution
The proposed solution is to allow the user to specify a global configuration for parameter serialization.

### Global config
Currently, spot doesn't provide a way to specify global configurations or overrides for default openapi rules, this proposal includes adding the class decorator **@config** which would be used in conjunction with the **@api** decorator. I'll discuss it's usage in the next sections.

### Specifying Parameters Serialization
To specify the parameter serialization rules, we would use an object for the key ** paramSerializationStrategy**, this object can have configurations for the serialization of `path`,  `query`, `header`, or `cookie`. 

The `query` configuration, will be composed by the keys `array` and `object`, and each of them have a limited set of possible values:

* **array:** `ampersand`, `ampersandEscaped`, `comma`, `pipe`
* **object:** `ampersand`, `comma`, `deep`

### Example
Here is an example of the configuration, the input, and the output.

**Config**
```ts
@api({ name: "My API" })
@config({
  paramSerializationStrategy: { 
    query: {
      array: "comma";
      object: "deep";
    }
  }
})
class MyApi {}
```

**Input**
```ts
// ...
request(
  @queryParams
  queryParams: {
    some_object: MyObject;
    some_array?: string[];
  }
  //...
) {}
```

**Output**
```yml
- in: query
  name: some_object
  required: true
  schema:
    $ref: '#...MyObject'
  style: deepObject # This was added due to the global rule "object: deep"
  explode: true # This was added due to the global rule "object: deep"

- in: query
  name: some_array
  required: false
  schema:
    type: array
    items:
       type: string
  style: form # This was added due to the global rule "array: comma"
  explode: false # This was added due to the global rule "array: comma"
```

**Comment by lfportal** [comment]
@alephao Spot will currently interpret that example as a `search` query parameter of type `object`, with properties `type` and `explode`.

It may be a reasonable (and good practice) to assume that APIs will be designed to follow a single convention for parameter serialisation. It would be convenient to be able to define the serialisation strategy globally.

**Comment by alephao** [comment]
@lfportal Thanks for clarifying! I updated the issue as per our conversation

Please review and let me know your thoughts

**Comment by lfportal** [comment]
@alephao the approach looks good. I'd rename `@configs` -> `@config` and consider `paramsSerialization` -> `paramSerializationStrategy`.

**Comment by alephao** [comment]
Sounds good, updated

**Comment by alephao** [comment]
Implemented the array strategies on #417
